zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+        .query_row(
+            "SELECT t.txid FROM ephemeral_addresses
+             LEFT OUTER JOIN transactions t
+             ON t.id_tx = COALESCE(used_in_tx, mined_in_tx)
+             WHERE address = :address",
+            named_params![":address": address_str],
+            |row| row.get::<_, Option<Vec<u8>>>(0),
+        )
+        .optional()?;
+
+    match res {
+        Some(Some(txid_bytes)) => {
+            let txid = TxId::from_bytes(
+                txid_bytes
+                    .try_into()
+                    .map_err(|_| SqliteClientError::CorruptedData("invalid txid".to_owned()))?,
+            );
+            Err(SqliteClientError::EphemeralAddressReuse(
+                address_str.to_owned(),
+                Some(txid),
+            ))
+        }
+        Some(None) if reject_all_ephemeral => Err(SqliteClientError::EphemeralAddressReuse(
+            address_str.to_owned(),
+            None,
+        )),
+        _ => Ok(()),
+    }
+}
+
+/// If `address` is one of our ephemeral addresses, mark it as having an output
+/// in a transaction that we have just created. This has no effect if `address` is
+/// not one of our ephemeral addresses.
+///
+/// Returns a `SqliteClientError::EphemeralAddressReuse` error if the address was
+/// already used.
+pub(crate) fn mark_ephemeral_address_as_used<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    ephemeral_address: &TransparentAddress,
+    tx_ref: i64,
+) -> Result<(), SqliteClientError> {
+    let address_str = encode_transparent_address_p(&wdb.params, ephemeral_address);
+    ephemeral_address_check_internal(wdb, &address_str, false)?;
+
+    wdb.conn.0.execute(
+        "UPDATE ephemeral_addresses SET used_in_tx = :used_in_tx WHERE address = :address",
+        named_params![":used_in_tx": &tx_ref, ":address": address_str],
+    )?;
+    Ok(())
+}
+
+/// If `address` is one of our ephemeral addresses, mark it as having an output
+/// in the given mined transaction (which may or may not be a transaction we sent).
+///
+/// `tx_ref` must be a valid transaction reference. This call has no effect if
+/// `address` is not one of our ephemeral addresses.
+pub(crate) fn mark_ephemeral_address_as_mined<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    address: &TransparentAddress,
+    tx_ref: i64,
+) -> Result<(), SqliteClientError> {
+    let address_str = encode_transparent_address_p(&wdb.params, address);
+
+    // Figure out which transaction was mined earlier: `tx_ref`, or any existing
+    // tx referenced by `mined_in_tx` for the given address. Prefer the existing
+    // reference in case of a tie or if both transactions are unmined.
+    // This slightly reduces the chance of unnecessarily reaching the gap limit
+    // too early in some corner cases (because the earlier transaction is less
+    // likely to be unmined).
+    //
+    // The query should always return a value if `tx_ref` is valid.
```

Comment:
The error message if this invariant doesn't hold is going to be inscrutable. Please use `.optional()` and return `CorruptedData` in the case that this doesn't return a row.

Reply:
That's not the right error, because it's more likely to be the caller providing an incorrect `tx_ref`. We only normally return `CorruptedData` in situations where we locally know that the database is storing something incorrect.

The argument for why the query must return a row when `tx_ref` is valid is that, since it's a left join, the `WHERE id_tx = :tx_ref` always matches if `tx_ref` is some `transactions(id_tx)`.

There doesn't seem to be any more appropriate `SqliteClientError` variant, so my inclination would be just to wrap the `rusqlite::Error` which would produce `DbError(rusqlite::Error::QueryReturnedNoRows)`.